File: plugins/jd/skills/error-audit/examples/test-projects/go-clean/payment/charge.go
package payment

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"runtime/debug"
)

var ErrChargeFailed = errors.New("charge failed")

func ChargeCustomer(ctx context.Context, customerID string, amount int) error {
	slog.Info("payment.charge.attempt",
		slog.String("customer_id", customerID),
		slog.Int("amount", amount),
	)

	if err := chargeProvider(amount); err != nil {
		return fmt.Errorf("charge customer %s: %w", customerID, err)
	}

	result, err := saveTransaction(customerID, amount)
	if err != nil {
		return fmt.Errorf("save transaction for %s: %w", customerID, err)
	}

	go func() {
		defer func() {
			if r := recover(); r != nil {
				slog.Error("notify.panic",
					slog.Any("recover", r),
					slog.String("stack", string(debug.Stack())),
				)
			}
		}()
		notifyCustomer(customerID, result)
	}()

	return nil
}

func chargeProvider(amount int) error                 { return nil }
func saveTransaction(id string, amt int) (any, error) { return nil, nil }
func notifyCustomer(id string, r any)                 {}

File: plugins/jd/skills/error-audit/examples/test-projects/nodejs-nestjs-clean/src/payment/charge.ts
import { PaymentError } from '../errors';
import { logger } from '../utils/logger';

export async function chargeCustomer(customerId: string, amount: number, paymentMethodToken: string) {
  try {
    const result = await chargeProvider.charge(amount, paymentMethodToken);
    return await saveTransaction(result);
  } catch (err) {
    logger.error('payment.charge_failed', { err, customerId, amount });
    throw new PaymentError('CHARGE_FAILED', 'Failed to charge customer', err as Error);
  }
}

declare const chargeProvider: any;
declare function saveTransaction(r: any): any;
